### DISCUSSION TITLE
Retention and Expiry confusion

### DISCUSSION THREAD

**Discussion by jonnepmyra** [minio/minio#18129] [Q&A] [answered]
Hello,

I am currently using MinIO version RELEASE.2023-08-29T23-07-35Z (go1.19.12 windows/amd64) hosted on a Windows environment. I have been working on achieving the automatic removal of objects once their specified retention time has elapsed. Each object is uploaded with a specific retention date, which is visible as metadata when using the MinIO object browser.

For instance, here is an example of such an object:

**Metadata**
```
Content-Type: application/xml
X-Amz-Object-Lock-Mode: GOVERNANCE
X-Amz-Object-Lock-Retain-Until-Date: 2023-09-27T00:00:06.000Z
```

**Object Info**
```
Legal Hold: Off
Retention Policy: Governance
```

The bucket where the object is uploaded is configured with two Lifecycle rules as follows:

```
Rule Type: Expire
Object Version: Current Version
Expiry days: 1
```

and

```
Rule Type: Expire
Object Version: Non-Current Version
Expiry days: 1
```
Despite this configuration, it appears that all objects are being removed, and the Retain-Until metadata does **not seem to be respected**. 

If I disable the Lifecycle rules, objects with a Retain-Until date that has passed are not being removed.

Is there something I might be overlooking, or is it not possible to have MinIO automatically remove objects when the Retain-Until date has passed?

I appreciate your assistance and thank you for your time.

**Comment by harshavardhana** [comment, marked as answer]
Don't use windows for production usecases.

**Reply by jonnepmyra** [reply]
Well, we're not using MinIO in a production environment at the moment; we're still in the early stages of development and evaluating its capabilities. However, I'm curious if the described configuration should achieve my goal of automatic removal, and objects with retention should not be removed if we were to host it on a Linux system instead?

**Reply by vadmeste** [reply]
> X-Amz-Object-Lock-Retain-Until-Date: 2023-09-27T00:00:06.000Z

@jonnepmyra this date is 16 hours ago, can you try again with further date ? just assuming there was an error in testing

**Comment by jonnepmyra** [comment]
> > X-Amz-Object-Lock-Retain-Until-Date: 2023-09-27T00:00:06.000Z
> 
> @jonnepmyra this date is 16 hours ago, can you try again with further date ?

Thanks for reaching out @vadmeste ! It was just an (bad)example, I have objects with future dates, up to a month or so..

They are all removed with the Expire Lifecycle rule enabled about a day after they've been uploaded.

Best regards

**Comment by vadmeste** [comment]
@jonnepmyra can you double check if objects are still there with 'mc ls --versions' ?

**Reply by jonnepmyra** [reply]
`mc ls --versions --recursive myminio/bucket/path/to/some/object/2023/09/27/00`

Shows alot of objects...  (for instance): 

```
[2023-09-29 08:11:58 CEST]     0B STANDARD 309cd753-0010-478e-b812-5d1bf67e1627 v2 DEL 00/018ad3b4-1639-3fa4-e25e-b7a362d1de6a
[2023-09-27 00:55:09 CEST]    95B STANDARD b92de866-57f3-47c4-8cac-1f053f74dd0b v1 PUT 00/018ad3b4-1639-3fa4-e25e-b7a362d1de6a
```

and if i toggle the "show deleted objects" in the MinIO Object Browser UI i can see the objects.

V2 of each object is [Deleted]
V1 of each object is the original

The object is however [Deleted] prematurely since the retention time has not elapsed.

When viewing the metadata of the V1 object in the UI, it no longer shows any metadata such as `X-Amz-Object-Lock-Retain-Until-Date`

All objects older than 1 day where deleted the same minute as i Enabled the Life Cycle Rule with 1 day Expire.

**Reply by harshavardhana** [reply]
yeah this is normal and expected behavior as per AWS S3 spect for object locking.

**Reply by harshavardhana** [reply]
DELETE marker is applied based on the time not the retention date, retention time is not a hint for ILM it is the object mtime. Retention time is kept for disallowing DELETEs from applications.
